File: backend/stub/com/nxshield/runtime/NXCrypto.java
package com.nxshield.runtime;

import java.io.ByteArrayOutputStream;
import java.security.MessageDigest;
import java.util.Arrays;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

/**
 * Mirrors the offline engine/crypto.py scheme:
 * MAGIC(4) || salt(8) || hmac-sha256/16 || rolling-xor(plain).
 */
public final class NXCrypto {

    private static final byte[] MAGIC = {'N', 'X', 'S', '1'};
    private static final int ROUNDS = 4096;

    private NXCrypto() {
    }

    public static byte[] decrypt(byte[] blob, byte[] password) {
        if (blob == null || blob.length < 28) {
            return new byte[0];
        }
        if (blob[0] != MAGIC[0] || blob[1] != MAGIC[1]
                || blob[2] != MAGIC[2] || blob[3] != MAGIC[3]) {
            throw new IllegalArgumentException("invalid nxshield blob");
        }
        byte[] salt = Arrays.copyOfRange(blob, 4, 12);
        byte[] digest = Arrays.copyOfRange(blob, 12, 28);
        byte[] xored = Arrays.copyOfRange(blob, 28, blob.length);
        byte[] dk = deriveKey(password, salt);
        byte[] check = hmac(dk, xored);
        if (!constantTimeEquals(Arrays.copyOfRange(check, 0, 16), digest)) {
            throw new SecurityException("blob mac mismatch");
        }
        return rollingXor(xored, dk);
    }

    public static byte[] deriveKey(byte[] password, byte[] salt) {
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            byte[] seed = new byte[password.length + salt.length];
            System.arraycopy(password, 0, seed, 0, password.length);
            System.arraycopy(salt, 0, seed, password.length, salt.length);
            byte[] dk = seed;
            for (int i = 0; i < ROUNDS; i++) {
                dk = md.digest(dk);
            }
            return dk;
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    public static byte[] rollingXor(byte[] data, byte[] key) {
        if (key == null || key.length == 0) {
            return data.clone();
        }
        byte[] out = new byte[data.length];
        int acc = 0xA5;
        for (int i = 0; i < data.length; i++) {
            int k = key[i % key.length] & 0xFF;
            acc = (acc + k + (i & 0xFF) + 1) & 0xFF;
            out[i] = (byte) (data[i] ^ k ^ (acc & 0xFF));
        }
        return out;
    }

    public static byte[] hmac(byte[] key, byte[] data) {
        try {
            Mac mac = Mac.getInstance("HmacSHA256");
            mac.init(new SecretKeySpec(key, "HmacSHA256"));
            return mac.doFinal(data);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    private static boolean constantTimeEquals(byte[] a, byte[] b) {
        if (a.length != b.length) {
            return false;
        }
        int r = 0;
        for (int i = 0; i < a.length; i++) {
            r |= a[i] ^ b[i];
        }
        return r == 0;
    }

    public static byte[] concat(byte[]... parts) {
        ByteArrayOutputStream bos = new ByteArrayOutputStream();
        for (byte[] p : parts) {
            bos.write(p, 0, p.length);
        }
        return bos.toByteArray();
    }
}

File: android/app/src/main/java/com/nxshield/engine/NxCrypto.kt
package com.nxshield.engine

import java.security.MessageDigest
import java.security.SecureRandom
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/** NXS1 安全封装：MAGIC + salt + HMAC 摘要 + 自逆滚动 XOR */
object NxCrypto {
    val MAGIC = "NXS1".toByteArray(Charsets.US_ASCII)

    private fun mode(b: Byte): Int = (b.toInt() and 0xFF) % 7 + 2

    /** 累加器滚动 XOR，只依赖 key/位置，自逆 */
    fun rollingXor(data: ByteArray, key: ByteArray): ByteArray {
        val out = ByteArray(data.size)
        var acc = 0
        for (i in data.indices) {
            val m = mode(key[i % key.size])
            acc = (acc + key[i % key.size].toInt()) and 0xFF
            out[i] = (((data[i].toInt() xor acc) and 0xFF) xor ((acc * m) and 0xFF)).toByte()
        }
        return out
    }

    fun sha256(b: ByteArray): ByteArray = MessageDigest.getInstance("SHA-256").digest(b)

    fun deriveKey(pass: ByteArray, salt: ByteArray): ByteArray {
        var k = sha256(pass + salt)
        repeat(8192) { k = sha256(k) }
        return k.copyOf(32)
    }

    fun hmac(key: ByteArray, data: ByteArray): ByteArray {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(sha256(key), "HmacSHA256"))
        return mac.doFinal(data)
    }

    /** 封装：[NXS1]+salt(8)+hmac(16)+payload(滚动XOR)。tag=HMAC  前16字节 */
    fun wrap(payload: ByteArray, password: ByteArray): ByteArray {
        val salt = ByteArray(8).also { SecureRandom().nextBytes(it) }
        val key = deriveKey(password, salt)
        val body = rollingXor(payload, key)
        val tag = hmac(key, MAGIC + salt + body).copyOf(16)
        return MAGIC + salt + tag + body
    }

    fun unwrap(data: ByteArray, password: ByteArray): ByteArray {
        require(data.size >= 8 + 8 + 16) { "NXS1: payload too short" }
        require(data.copyOf(4).contentEquals(MAGIC)) { "NXS1: bad magic" }
        val salt = data.copyOfRange(4, 12)
        val tag = data.copyOfRange(12, 28)
        val body = data.copyOfRange(28, data.size)
        val key = deriveKey(password, salt)
        val expect = hmac(key, MAGIC + salt + body).copyOf(16)
        require(MessageDigest.isEqual(expect, tag)) { "NXS1: HMAC mismatch (wrong key / tampered)" }
        return rollingXor(body, key)
    }
}